File: ts/core/scalar_quant.ts
/**
 * Scalar Quantization Module for TurboQuant
 * 
 * Implements MSE-optimal Lloyd-Max scalar quantization with random rotation.
 * The rotation ensures coordinates follow a concentrated distribution (approx. Gaussian).
 */

import * as tf from '@tensorflow/tfjs';

export interface ScalarQuantizedData {
  indices: tf.Tensor;
  scales: tf.Tensor;
  norms: tf.Tensor;
  rotationMatrix: tf.Tensor2D;
}

export class ScalarQuantizer {
  private static codebookCache: Map<number, { centroids: tf.Tensor1D, boundaries: tf.Tensor1D }> = new Map();

  /**
   * Generate a random orthogonal matrix via QR decomposition.
   */
  public static generateRotationMatrix(d: number, seed: number = 42): tf.Tensor2D {
    return tf.tidy(() => {
      // Generate random Gaussian matrix
      const A = tf.randomNormal([d, d], 0, 1, 'float32', seed);
      
      // QR decomposition to get orthogonal matrix Q
      const [q, r] = tf.linalg.qr(A);
      return q as tf.Tensor2D;
    });
  }

  /**
   * Compute Lloyd-Max optimal quantization codebook for N(0, 1).
   */
  public static async getCodebook(numBits: number): Promise<{ centroids: tf.Tensor1D, boundaries: tf.Tensor1D }> {
    if (this.codebookCache.has(numBits)) {
      return this.codebookCache.get(numBits)!;
    }

    const numLevels = Math.pow(2, numBits);
    const numSamples = 100000;

    const result = tf.tidy(() => {
      // Sample from standard normal
      const samples = tf.randomNormal([numSamples], 0, 1);
      const sampleValues = Array.from(samples.dataSync());
      
      // Initialize centroids uniformly in [-4, 4]
      let centroidValues = Array.from(tf.linspace(-4, 4, numLevels).dataSync());
      
      // Iterative Lloyd-Max optimization (simplified for implementation)
      // For production, precomputing these is better.
      for (let i = 0; i < 20; i++) {
        const sums = new Array<number>(numLevels).fill(0);
        const counts = new Array<number>(numLevels).fill(0);

        for (const sample of sampleValues) {
          let bestLevel = 0;
          let bestDistance = Number.POSITIVE_INFINITY;

          for (let level = 0; level < numLevels; level++) {
            const distance = Math.abs(sample - centroidValues[level]);
            if (distance < bestDistance) {
              bestDistance = distance;
              bestLevel = level;
            }
          }

          sums[bestLevel] += sample;
          counts[bestLevel] += 1;
        }

        const newCentroidsArr: number[] = [];
        for (let l = 0; l < numLevels; l++) {
          if (counts[l] > 0) {
            newCentroidsArr.push(sums[l] / counts[l]);
          } else {
            newCentroidsArr.push(centroidValues[l]);
          }
        }
        centroidValues = newCentroidsArr;
      }
      
      const boundariesArr: number[] = [];
      for (let i = 0; i < centroidValues.length - 1; i++) {
        boundariesArr.push((centroidValues[i] + centroidValues[i + 1]) / 2);
      }
      
      return {
        centroids: tf.tensor1d(centroidValues),
        boundaries: tf.tensor1d(boundariesArr)
      };
    });

    this.codebookCache.set(numBits, result);
    return result;
  }

  /**
   * Quantize input vectors.
   */
  public static async quantize(
    x: tf.Tensor,
    numBits: number,
    rotationSeed: number = 42
  ): Promise<ScalarQuantizedData> {
    const d = x.shape[x.shape.length - 1];
    const rotationMatrix = this.generateRotationMatrix(d, rotationSeed);
    
    return tf.tidy(() => {
      const xFlat = x.reshape([-1, d]);
      
      // 1. Rotate
      const xRotated = tf.matMul(xFlat, rotationMatrix);
      
      // 2. Compute norms and scales
      const norms = tf.norm(xRotated, 2, 1, true);
      const scales = tf.div(norms, Math.sqrt(d));
      
      // 3. Normalize
      const xNormalized = tf.div(xRotated, tf.add(scales, 1e-8));
      
      // 4. Quantize using precomputed codebook
      // Note: In real TFJS, we'd use a more optimized way to digitize
      // For now, we'll do a simple comparison
      const { centroids, boundaries } = this.codebookCache.get(numBits) || { centroids: tf.tensor1d([]), boundaries: tf.tensor1d([]) };
      
      // Simple bucketization
      const indices = tf.sum(tf.cast(tf.greater(xNormalized.expandDims(-1), boundaries.expandDims(0).expandDims(0)), 'int32'), -1);
      
      return {
        indices: indices.reshape(x.shape),
        scales: scales.reshape([...x.shape.slice(0, -1), 1]),
        norms: norms.reshape([...x.shape.slice(0, -1), 1]),
        rotationMatrix
      };
    });
  }

  /**
   * Dequantize indices back to vectors.
   */
  public static dequantize(
    data: ScalarQuantizedData,
    numBits: number
  ): tf.Tensor {
    const { centroids } = this.codebookCache.get(numBits)!;
    
    return tf.tidy(() => {
      const d = data.rotationMatrix.shape[0];
      const indicesFlat = data.indices.reshape([-1, d]);
      const scalesFlat = data.scales.reshape([-1, 1]);
      
      // 1. Lookup centroids
      const xNormalized = tf.gather(centroids, indicesFlat);
      
      // 2. Rescale
      const xScaled = tf.mul(xNormalized, scalesFlat);
      
      // 3. Inverse rotation (Transpose of orthogonal matrix)
      const xRotated = tf.matMul(xScaled, data.rotationMatrix, false, true);
      
      return xRotated.reshape(data.indices.shape);
    });
  }
}

File: core/scalar_quant.py
"""
Scalar Quantization Module for TurboQuant

Implements MSE-optimal Lloyd-Max scalar quantization with random rotation.
The rotation ensures coordinates follow a concentrated distribution (approx. Gaussian),
making scalar quantization near-optimal.
"""

import torch
from torch import Tensor
from typing import Tuple, Optional
import math
from .wht import apply_random_wht


def _generate_rotation_matrix(d: int, seed: int = 42, device: Optional[torch.device] = None) -> Tensor:
    """
    Generate a random orthogonal matrix via QR decomposition of Gaussian matrix.
    """
    generator = torch.Generator(device=device)
    generator.manual_seed(seed)
    A = torch.randn(d, d, generator=generator, device=device)
    Q, _ = torch.linalg.qr(A)
    return Q


def _compute_lloyd_max_codebook(
    num_bits: int,
    num_samples: int = 100000,
    seed: int = 42,
    device: Optional[torch.device] = None
) -> Tuple[Tensor, Tensor]:
    """
    Compute Lloyd-Max optimal quantization codebook via numerical optimization.
    """
    num_levels = 2 ** num_bits
    generator = torch.Generator(device=device)
    generator.manual_seed(seed)
    samples = torch.randn(num_samples, generator=generator, device=device)
    
    centroids = torch.linspace(-4, 4, num_levels + 1, device=device)
    centroids = (centroids[:-1] + centroids[1:]) / 2
    
    for _ in range(50):
        distances = (samples.unsqueeze(1) - centroids.unsqueeze(0)).abs()
        assignments = distances.argmin(dim=1)
        
        counts = torch.zeros(num_levels, device=device)
        ones = torch.ones_like(assignments, dtype=torch.float32)
        counts = counts.scatter_add(0, assignments, ones)
        
        sums = torch.zeros(num_levels, device=device)
        sums = sums.scatter_add(0, assignments, samples)
        
        mask = (counts > 0)
        new_centroids = centroids.clone()
        new_centroids[mask] = sums[mask] / counts[mask]
        centroids = new_centroids
    
    boundaries = (centroids[:-1] + centroids[1:]) / 2
    return centroids, boundaries


_codebook_cache: dict = {}


def get_codebook(
    num_bits: int,
    device: Optional[torch.device] = None
) -> Tuple[Tensor, Tensor]:
    """
    Get or compute Lloyd-Max codebook for given bit width.
    """
    cache_key = (num_bits, str(device) if device else 'cpu')
    if cache_key not in _codebook_cache:
        centroids, boundaries = _compute_lloyd_max_codebook(num_bits, device=device)
        _codebook_cache[cache_key] = (centroids, boundaries)
    return _codebook_cache[cache_key]


def quantize_scalar(
    x: Tensor,
    num_bits: int,
    rotation_matrix: Optional[Tensor] = None,
    rotation_seed: int = 42,
    rotation_type: str = "random",
    eps: float = 1e-8
) -> Tuple[Tensor, Tensor, Tensor, Optional[Tensor]]:
    """
    Apply MSE-optimal scalar quantization to input vectors.
    """
    orig_shape = x.shape
    device = x.device
    x_flat = x.view(-1, x.shape[-1])
    N, d = x_flat.shape
    
    if rotation_type == "hadamard":
        x_rotated = apply_random_wht(x_flat, seed=rotation_seed)
    else:
        if rotation_matrix is None:
            rotation_matrix = _generate_rotation_matrix(d, rotation_seed, device)
        x_rotated = x_flat @ rotation_matrix
    
    norms = x_rotated.norm(dim=1, keepdim=True)
    is_zero = norms < eps
    scales = torch.where(is_zero, torch.ones_like(norms), norms / math.sqrt(d))
    x_normalized = x_rotated / (scales + eps)
    
    centroids, boundaries = get_codebook(num_bits, device)
    indices = torch.searchsorted(boundaries, x_normalized.contiguous())
    indices = indices.clamp(0, len(centroids) - 1)
    
    indices = indices.view(orig_shape)
    scales = scales.view(*orig_shape[:-1], 1)
    norms = norms.view(*orig_shape[:-1], 1)
    
    return indices, scales, norms, rotation_matrix


def dequantize_scalar(
    indices: Tensor,
    scales: Tensor,
    num_bits: int,
    rotation_matrix: Optional[Tensor] = None,
    rotation_seed: int = 42,
    rotation_type: str = "random"
) -> Tensor:
    """
    Reconstruct vectors from scalar quantization indices.
    """
    orig_shape = indices.shape
    device = indices.device
    indices_flat = indices.view(-1, indices.shape[-1])
    scales_flat = scales.view(-1, 1)
    
    centroids, _ = get_codebook(num_bits, device)
    x_normalized = centroids[indices_flat]
    x_scaled = x_normalized * scales_flat
    
    if rotation_type == "hadamard":
        x_hat = apply_random_wht(x_scaled, seed=rotation_seed)
    else:
        if rotation_matrix is None:
            rotation_matrix = _generate_rotation_matrix(indices_flat.shape[-1], rotation_seed, device)
        x_hat = x_scaled @ rotation_matrix.T
    
    return x_hat.view(orig_shape)


def quantize_and_reconstruct(
    x: Tensor,
    num_bits: int,
    rotation_seed: int = 42
) -> Tuple[Tensor, dict]:
    """
    Convenience function for quantization and immediate reconstruction.
    """
    indices, scales, norms, rotation_matrix = quantize_scalar(
        x, num_bits, rotation_seed=rotation_seed
    )
    x_hat = dequantize_scalar(indices, scales, num_bits, rotation_matrix)
    metadata = {'indices': indices, 'scales': scales, 'norms': norms, 'rotation_matrix': rotation_matrix}
    return x_hat, metadata
